client/landing/stepper/README.md

```diff
@@ -1,106 +1,402 @@
+
+They should **only submit** to the flow, and the flow should do all the thinking, persisting, and the navigation.
+
+**Common question**: What if a step wants to cancel/skip/drop out? Then it should submit something like `{ action: skip }`.
+
+<details><summary><strong>Expand if you're curious why.</strong></summary>
+<p>
+	<ol>
+		<li>When steps make navigational decisions, the aforementioned finite-state machine is thrown out of the window. The dependency graph explodes and things get out of control quickly. Especially if you consider that steps are reusable across flows.</li>
+		<li>Same thing happens when steps communicate: the flow is blind to some decisions, and the steps become less re-usable.</li>
+		<li>When a step has side effects in Flow A, they might affect the step's behaviour in Flow B. Only the flow should persist things.</li>
+	</ol>
+</p>
+</details>
+
+In general, the smarter the step, the more problematic and tailored it is. Please treat your steps as buttons or inputs.
+
+### Making a flow
+
+A flow is a collection of steps. Each of these steps submit some information to the flow, which means the state of the flow is largely the sum of these submitted data. For that reason, the `FlowV2` interface requires the steps collection to be defined before the flow itself. This way, the flow can shape its state around the submissions and properties of these steps.
+
+**Note:** We have an example flow you can use as a reference [here](/client/landing/stepper/declarative-flow/flows/00-example-flow/example.ts).
+
+#### Code example
 
 ```ts
-import type { StepPath } from './internals/steps-repository';
-import type { Flow } from './internals/types';
+import { useFlowState } from '../../internals/state-manager/store';
+import { STEPS } from '../../internals/steps';
 
-export const exampleFlow: Flow = {
-	initialize() {
-		return [ STEPS.DOMAINS, STEPS.DESIGN ];
-	},
-	useStepNavigation( currentStep, navigate ) {
-		const goBack = () => {
-			if ( currentStep === 'domain' ) {
-				navigate( 'design' );
-			} else {
-				navigate( 'domain' );
+// We define our steps collection upfront because it is a prerequisite of shaping the flow.
+async function initialize( calypsoReduxStore: Store ) {
+	const includeDomainsStep = await isTheMoonFull();
+	const hasAnySites = userHasAnySites( calypsoReduxStore.getState() );
+
+	if ( includeDomainsStep ) {
+		return [ STEPS.UNIFIED_DOMAINS, STEPS.UNIFIED_PLANS, STEPS.PROCESSING, STEPS.ERROR ] as const;
+	}
+
+	if ( hasAnySites ) {
+		return [ STEPS.PICK_SITE, STEPS.UNIFIED_DOMAINS, STEPS.UNIFIED_PLANS, STEPS.PROCESSING, STEPS.ERROR ] as const;
+	}
+
+	// We need `as const` to promise TS that these steps won't change later.
+	return [ STEPS.UNIFIED_PLANS, STEPS.PROCESSING, STEPS.ERROR ] as const;
+}
+
+export const exampleFlow: FlowV2< typeof initialize > = {
+	/**
+	 * The name of the flow is what appears in the pathname. It must be unique.
+	 * This flow will be under /setup/my-flow.
+	 */
```

Comment:
A word of caution: code comments are great, but they can quickly get out of date with the TypeScript JSDocs.